### ISSUE TITLE
Overall suggestions for UI/UX & feature improvements

### ISSUE THREAD

**Issue by Cyberavater** [MuntashirAkon/AppManager#347]
- [x] Add "Unblock trackers" just as we have "Block trackers" in both batch and per-app menu.
- [x] Add Select All & Unselect all button in 1-Click Ops's "Block/unblock trackers", "Backup", "Restore" menu and in Backup/Restore menu [This could be a checkbox at the top left, marking it would select all and un-marking it would un-select all.]
- [x] U̷p̷d̷a̷t̷e̷ ̷a̷p̷p̷ ̷v̷e̷r̷s̷i̷o̷n̷ ̷c̷o̷d̷e̷ ̷p̷e̷r̷ ̷d̷e̷b̷u̷g̷ ̷u̷p̷d̷a̷t̷e̷,̷ ̷t̷h̷i̷s̷ ̷s̷h̷o̷u̷l̷d̷ ̷b̷e̷ ̷m̷o̷r̷e̷ ̷p̷r̷a̷c̷t̷i̷c̷a̷l̷ ̷i̷n̷ ̷d̷e̷b̷u̷g̷g̷i̷n̷g̷ ̷t̷h̷a̷n̷ ̷t̷h̷e̷ ̷c̷u̷r̷r̷e̷n̷t̷ ̷w̷o̷r̷k̷f̷l̷o̷w̷.̷ ̷N̷o̷ ̷a̷d̷v̷e̷r̷t̷ ̷b̷u̷t̷ ̷a̷ ̷f̷i̷n̷e̷ ̷e̷x̷a̷m̷p̷l̷e̷ ̷w̷o̷u̷l̷d̷ ̷b̷e̷ ̷h̷t̷t̷p̷s̷:̷/̷/̷t̷.̷m̷e̷/̷L̷S̷P̷o̷s̷e̷d̷A̷r̷c̷h̷i̷v̷e̷s̷ - Apart from bad practice, changing version code is a bad idea because the user wouldn't be able to downgrade in the event of the app being too unstable. Instead, the run number is included in the version name string (cc5fb88).
- [x] Rename Backup APK option to Extract APK(s) (or Export APK maybe?) [Since we already have a Backup feature, this would keep things more clear] [Also, this Backup APK feature is pretty limited it doesn't add the .obb file]
- [x] Add .obb file when exporting apks. [If .apks doesn't support .obb file then why not create a new .apks version under AM that supports OBB? If that's not feasible maybe switch to .xapk or make a whole new format under AM.]
- [x] Separate the UI for Instructions/Docs (As we have for "Log Viewer"). [This would help the user experiment with AM while reading the docs]
- [x] Add a search bar in Instructions/Docs [This would help the user to quickly find what he is looking for or even if such a thing exists in AM or not.]
- [x] Merge Enable and disable option. [As you've done with Block and unblock trackers. Also, check if this option list requires reordering.]
- [x] List options - Filter - Keystore apps
- [ ] Inverse Filtering? [IDK if it's feasible, just strikethrough it, if impossible]
- [x] Instantly commit _specific_ changes to the database when the user makes changes using AM (i.e. Backup, Uninstall, Disable, etc). Refreshing the whole Onboard page works but it takes time, as one of the reasons to implement the database was to improve load time, changes to the database should be committed immediately even without refreshing.
- [ ] After both thread safety is ensured and delta backup is implemented, add a progress bar in the notification for Backup & Restore, that progresses per components, rather than per apps.
- [ ] Replace "Prevent background operation" with Prevent/Restore background operation [Like we have for trackers].
- [x] If something is selected and I press the back button the app closes, this is actually not expected. You see, most if not all apps, unselect the items on first back press and when all app is unselected only then the back button is actually used for going to the previous screen. Eg apps: Mix, Telegram. So, this app should also be consistent with users' habits.
- [x] The current popup section for selected apps really leaves little space for apps to be selected if the keyboard is up. So, maybe it was not really a wise decision to change UI like this? If you really like this UI, maybe I can suggest some improvements? I.e. first let the user select all the apps he wants to select, then when the user's selection is complete, the user would press a FloatingActionButton, which would be available on the bottom right. Now, this would pop up your newly implemented window in the center with more space (almost like the current Backup/Restore screen).
- [ ] Running apps - Add CPU usage (time + percent) information [Add CPU usage in 'sort by' too] [Add filters, i.e, hide system apps]

<details>
<summary>Moved some suggestion to #355 upon the dev's request</summary>
<br>
Considerable suggestion (since given below will override existing UI/UX and function):

- If the backed-up app version is lower than the installed version, "Backup" should be written in "Red" vice-versa if the backed-up app version is equal or higher, "Backup" should be written in "Green". [Also, making the text bold should look nice]. This would require another way to show the apps that are Backup up but uninstalled. One (my own) suggestion is to make them faded in the list.
- "Data" should not be selected for backing up "External Data". The user could be interested in only external data but not data. (Example case; The user doesn't want to back up the progress of a game but wants external data as he doesn't want to re-download the large external data that he already downloaded once.
-  Make Backup/Restore menu options more Syntacticly Sound. For example, the option that deals with Cache should be named "Cache", selecting it would include the app's cache in the backup. This change would impact and remodel the following;
1. APK (+Split) [This would include apps apk and split-apk in backup/restore]
2. ODEX/VDEX [This would include apps odex and vdex in backup/restore], [The dev can name it something else, i.e. "Regeneratives" (as it regenerates per install/updates,) and specify in the description what it includes] [This option is still questionable, as I've yet to see it's usefulness. #313]
3. "Cache" [Of course! as explained above]

Of course, this would re-format the current order and would put the "ODEX/VDEX" option just above the "Cache" option.
</details>

**Comment by MuntashirAkon** [comment]
> Update app version code per debug update

Apart from bad practice, changing version code is a bad idea because the user wouldn't be able to downgrade in the event of the app being too unstable. Instead, the run number is included in the version name string (cc5fb88a).

**Comment by Cyberavater** [comment]
You can StrikeThrough m̷y̷ ̷s̷u̷g̷g̷e̷s̷t̷i̷o̷n̷ in the OP and Type your own final decision and mark it as done for convenience.

**Comment by MuntashirAkon** [comment]
@Cyberavater Create another issue with the last three suggestions as they are related to back up/restore and should have higher priorities than the above ones.

**Comment by Cyberavater** [comment]
OP updated accordingly and with a suggestion regarding mentioned to tackle implement the introduced function.

**Comment by Cyberavater** [comment]
@MuntashirAkon What does delete backup for user0 mean? Shouldn't it be deleted for all users? Can you update dos regarding this for clarification?

**Comment by Cyberavater** [comment]
OP updated with more suggestions, title updated accordingly.

**Comment by MuntashirAkon** [comment]
> If that's not feasible maybe switch to .xapk or make a whole new format under AM.

Yes, I'm working on it. See #358.

**Comment by Cyberavater** [comment]
Some more suggestions added.

**Comment by gnuhead-chieb** [comment]
It should be show backup/restore options as horizontally

https://user-images.githubusercontent.com/41156994/115986585-6737e980-a5a0-11eb-9f01-f82cc8ef7b3f.mp4

**Comment by MuntashirAkon** [comment]
> It should be show backup/restore options as horizontally

It depends on the length of the strings. If you reduce them, they'll be displayed horizontally, in a single line. If you have better ideas, consider opening an issue in the [material components](https://github.com/material-components/material-components-android) repo because it's their technology.

**Comment by gnuhead-chieb** [comment]
> It depends on the length of the strings. If you reduce them, they'll be displayed horizontally

OK,I understand :)

**Comment by MuntashirAkon** [comment]
Don't worry much about it as the current back up/restore UI will be changed in a future release.

**Comment by gnuhead-chieb** [comment]
I also have UI suggestion.
In class viewer at Scanner,there is a button named "Toggle Class Listing" for switch between showing all classes or showing only tracker's classes.

It should be separate as "Show tracker's class" and "Show all class"

**Comment by MuntashirAkon** [comment]
> It should be separate as "Show tracker's class" and "Show all class"

Maybe replacing it with an icon would be better so that a user can toggle between the class lists easily.

**Comment by Cyberavater** [comment]
@MuntashirAkon I've added some more suggestions, please kindly check.

**Comment by MuntashirAkon** [comment]
> The current popup section for selected apps really leaves little space for apps to be selected if the keyboard is up. So, maybe it was not really a wise decision to change UI like this? If you really like this UI, maybe I can suggest some improvements? I.e. first let the user select all the apps he wants to select, then when the user's selection is complete, the user would press a FloatingActionButton, which would be available on the bottom right. Now, this would pop up your newly implemented window in the center with more space (almost like the current Backup/Restore screen).

No, that's too much workflow. The purpose of changing the selection panel is to reduce the amount of workflow as the previous workflow was very tedious for certain users including Android TV users. The action bar will also be replaced gradually in the same way which is also unusable for devices lacking touch support. When it is done, I might consider hiding the bottom section of the selection pane when a keyboard is activated and there's little space left for selecting apps (as it depends on the height/DPI of the device).

**Comment by Cyberavater** [comment]
Not to argue, but I don't think 1 tap more is that "too much workflow". And, I hope you do realize that your mentioned exotic devices don't actually cover the majority. 
Anyhow, you're the boss. But, can I at least request the previous UI, which will be enabled via a setting? No offense, but currently the app is really unusable, IMHO.

**Comment by Cyberavater** [comment]
After selecting 60 apps (spending 10 mins) to backup, I didn't find the Backup option (maybe due to one app selected that's not installed), although I suggested this, but seems more harm than good.

Can you revert this change? i.e show the Backup button even when one app is not installed?

**Comment by gnuhead-chieb** [comment]
It should be openable app info page from installer

Here is concept image made with gimp
![Screenshot_20210910-115424](https://user-images.githubusercontent.com/41156994/132793972-d8023cfa-da04-4dd9-9a4a-798c33f49477.png)

**Comment by MuntashirAkon** [comment]
> It should be also openable app info page from installer before instalation

I'm still investigating whether I should add direct link to scanner or app details page there.

**Comment by Cyberavater** [comment]
Add an option in settings to keep the installation progress in the background by default.

**Comment by MuntashirAkon** [comment]
> Add an option in settings to keep the installation progress in the background by default.

Implemented in 218369f6.

**Comment by MuntashirAkon** [comment]
>  After both thread safety is ensured and delta backup is implemented, add a progress bar in the notification for Backup & Restore, that progresses per components, rather than per apps.

What do you mean by _components_? Can you elaborate on this point? Also, what's wrong with per app progress (I understand that it's useless if the user is only backing up a single app)?

**Comment by MuntashirAkon** [comment]
>  After both thread safety is ensured and delta backup is implemented, add a progress bar in the notification for Backup & Restore, that progresses per components, rather than per apps.

What do you mean by _components_? Can you elaborate on this point? Also, what's wrong with per app progress (I understand that it's useless if the user is only backing up a single app)?

**Comment by Cyberavater** [comment]
> What do you mean by _components_? 

Assume, the user is backing up an app with apk + internal data. So, backing up apk would be 50% task complete, and with internal data backed up, progress will reach 100%. True progress bar by size will be more preferable, but I don't know if that's feasible.

**Comment by MuntashirAkon** [comment]
> True, the progress bar by size will be more preferable, but I don't know if that's feasible.

It might be possible by calculating size of the backup before starting the backup process. Anyway, this wasn't the answer I was looking for. I was rather interested in a rather elaborate answer as it would be very easy for me to understand how the progress bar should work and set the logic to match that expectation.

**Comment by Cyberavater** [comment]
> > Not to argue, but I don't think 1 tap more is that "too much workflow". And, I hope you do realize that your mentioned exotic devices don't actually cover the majority.
> 
> Yes, they don't cover the majority. But you're again forgetting the fact that App Manager isn't a product, it's a hobby project and its design choices depend solely on my choice and my use-cases. What you and others can only provide are opinions and if they're sane and match my tastes, I might implement them or at least think positively about them. It might sound a little offensive but this is how it works. If the project doesn't match my tastes and use-case, I might loose interest in using it which also means the death of the project. Now, this won't be true if I managed to get a good funding for it but I guess I would have to wait a few more years to manage that.
> 
> That being said, [59884c3](https://github.com/MuntashirAkon/AppManager/commit/59884c3df82324a5d79c94e8a7bca33962b8bd6e) has solved this issue (read the commit message for details).

Well, thanks for your continued work. Also, forgive me if I sounded rude.

>  I was rather interested in an elaborate answer as it would be very easy for me to understand how the progress bar should work and set the logic to match that expectation.

By component:

Assume 2 apps, one with app + internal data and the other let's say only internal data.

Now, If I start backup (considering one by one i.e no parallel), 33% is done when the first app backup completes, 66% is done when internal data is done, now after internal data backup is done for the 2nd app 100% completes.
[This method is used by SB]

By size:

Assume 2 apps, one with x+y=z1 (z would be the size of the backup) and the other with a = z2. Now, total progress would be z1+z2=100%. The amount of data that has been backed up (or restored) would be the current percentage. Again, I don't know if it's feasible, i.e, can we tell the exact amount of data that has been restored at a given time?

**Comment by MuntashirAkon** [comment]
> List options - Filter - Keystore apps

Since App Manager now supports restoring apps with KeyStore. This filter is no longer necessary.

**Comment by MuntashirAkon** [comment]
> Running apps - Add CPU usage (time + percent) information [Add CPU usage in 'sort by' too] [Add filters, i.e, hide system apps]

This won't be implemented because **CPU usage** does not mean what people think it means and will give a false sense of how CPU is used.
